### ISSUE TITLE
CUDA exposed in interface

### ISSUE THREAD

**Issue by jeanlucf22** [hypre-space/hypre#84]
(Sorry for posting previously to the wrong github. Here is an update using the latest
version of hypre downloaded today from https://github.com/hypre-space/hypre)

I am using the hypre struct PFMG solver built for GPU on the Volta V 100.
I built hypre with the options:
./configure --with-cuda --enable-unified-memory --enable-shared
--prefix=$HOME/hypre/summit_xl_gpu CUDACXX=mpiCC HYPRE_CUDA_SM=70

While it is working for me, I had to work around an interface issue:
To define various hypre structure in my C++ code, such as HYPRE_StructVector that
I can pass to initialization functions such as HYPRE_StructVectorCreate(), I have to include
the file HYPRE_struct_ls.h.

Now if I use a C++ compiler (xlC here), I end up with an error like:

In file included from /ccs/home/jeanluc/GIT/PFiSM/source/PfmFACSolver.C:2:
In file included from /ccs/home/jeanluc/GIT/PFiSM/source/CellPoissonHypreSolver.h:25:
In file included from /ccs/home/jeanluc/hypre/summit_xl_gpu/include/HYPRE_struct_ls.h:1105:
In file included from /ccs/home/jeanluc/hypre/summit_xl_gpu/include/interpreter.h:11:
/ccs/home/jeanluc/hypre/summit_xl_gpu/include/_hypre_utilities.h:1260:18: error: use of undeclared
      identifier 'blockDim'
         return (blockDim.x);

In other words, even though I have no CUDA code in my own code, I end up having to
compile multiple C++ files with nvcc. I was wondering why CUDA code/includes in hypre are
exposed in the interface and lead to that issue? I noticed a comment in HYPRE_struct_ls.h
"These includes shouldn't be here. (RDF)", so I am guessing that that issue is already
on the developers radar? Or maybe I am missing a configure option?

Thanks,
Jean-Luc Fattebert
ORNL

**Comment by liruipeng** [comment]
@rfalgout @ulrikeyang @oseikuffuor1 I see what the problem is; we should talk about this at some time

**Comment by rfalgout** [comment]
Hi @jeanlucf22 .  Based on the comment you mention, this issue was indeed once on the radar, but it has long dropped off. :)  I will try to clean this up now.  In general, no lower-case '\_hypre_\*.h' header files should be included by way of capital 'HYPRE_\*.h' header files.  Thanks for letting us know!

**Comment by rfalgout** [comment]
Hi @jeanlucf22 .  Pull request #86 should fix this issue once it's merged.

**Comment by rfalgout** [comment]
Hi @jeanlucf22 .  This issue should be fixed now on master.  Let me know if that is not the case.  Thanks!

**Comment by jeanlucf22** [comment]
I think the issue is still there, though through a different path:

In file included from /ccs/home/jeanluc/GIT/PFiSM/source/main.C:35:
In file included from /ccs/home/jeanluc/hypre/summit_xl_gpu/include/HYPRE_struct_ls.h:13:
In file included from /ccs/home/jeanluc/hypre/summit_xl_gpu/include/HYPRE_lobpcg.h:13:
In file included from /ccs/home/jeanluc/hypre/summit_xl_gpu/include/fortran_matrix.h:11:
/ccs/home/jeanluc/hypre/summit_xl_gpu/include/_hypre_utilities.h:1260:18: error: use of undeclared identifier
      'blockDim'
         return (blockDim.x);
                 ^
/ccs/home/jeanluc/hypre/summit_xl_gpu/include/_hypre_utilities.h:1262:18: error: use of undeclared identifier
      'blockDim'
         return (blockDim.x * blockDim.y);

**Comment by rfalgout** [comment]
Hi @jeanlucf22 .  I'll work on getting a fix for this.  In general, I think we need a way to test and regression test include file usage in hypre so that we can prevent this sort of thing from creeping into the library.  We are currently even abusing header usage in our example codes!

Stay tuned.  Thanks!

**Comment by rfalgout** [comment]
Hi @jeanlucf22 .  When you get a chance, would you give the master branch another try.  Hopefully the CUDA compile issue is all cleared up now.  Let me know.  Thanks!

**Comment by jeanlucf22** [comment]
Hi @rfalgout . I just gave the master branch a try, and it now all works for me. Much cleaner code on my side now. Thanks!

**Comment by rfalgout** [comment]
Excellent!  Thanks for taking a look @jeanlucf22 !
